sdk/src/types/block/semantic.rs

```diff
@@ -209,294 +210,279 @@ impl<'a> ValidationContext<'a> {
+                        .checked_add(native_token.amount())
+                        .ok_or(Error::CreatedNativeTokensAmountOverflow)?;
                 }
             }
         }
 
-        if let Some(sender) = features.and_then(|f| f.sender()) {
-            if !context.unlocked_addresses.contains(sender.address()) {
-                return Ok(Some(TransactionFailureReason::SenderNotUnlocked));
+        // Validation of storage deposit returns.
+        for (return_address, return_amount) in self.storage_deposit_returns.iter() {
+            if let Some(deposit_amount) = self.simple_deposits.get(return_address) {
+                if deposit_amount < return_amount {
+                    return Ok(Some(TransactionFailureReason::StorageDepositReturnUnfulfilled));
+                }
+            } else {
+                return Ok(Some(TransactionFailureReason::StorageDepositReturnUnfulfilled));
             }
         }
 
-        context.output_amount = context
-            .output_amount
-            .checked_add(amount)
-            .ok_or(Error::CreatedAmountOverflow)?;
-
-        context.output_mana = context
-            .output_mana
-            .checked_add(mana)
-            .ok_or(Error::CreatedManaOverflow)?;
-
-        if let Some(created_native_tokens) = created_native_tokens {
-            for native_token in created_native_tokens.iter() {
-                let native_token_amount = context
-                    .output_native_tokens
-                    .entry(*native_token.token_id())
-                    .or_default();
-
-                *native_token_amount = native_token_amount
-                    .checked_add(native_token.amount())
-                    .ok_or(Error::CreatedNativeTokensAmountOverflow)?;
-            }
+        // Validation of amounts.
+        if self.input_amount != self.output_amount {
+            return Ok(Some(TransactionFailureReason::SumInputsOutputsAmountMismatch));
         }
-    }
 
-    // Validation of storage deposit returns.
-    for (return_address, return_amount) in context.storage_deposit_returns.iter() {
-        if let Some(deposit_amount) = context.simple_deposits.get(return_address) {
-            if deposit_amount < return_amount {
-                return Ok(Some(TransactionFailureReason::StorageDepositReturnUnfulfilled));
-            }
-        } else {
-            return Ok(Some(TransactionFailureReason::StorageDepositReturnUnfulfilled));
+        if self.input_mana > self.output_mana && !self.transaction.has_capability(TransactionCapabilityFlag::BurnMana) {
+            // TODO: add a variant https://github.com/iotaledger/iota-sdk/issues/1430
+            return Ok(Some(TransactionFailureReason::SemanticValidationFailed));
         }
-    }
 
-    // Validation of amounts.
-    if context.input_amount != context.output_amount {
-        return Ok(Some(TransactionFailureReason::SumInputsOutputsAmountMismatch));
-    }
+        let mut native_token_ids = HashSet::new();
```

Comment:
since we know the required capacity maybe we should make use of that?

Reply:
DOne